generate_ci.py

```diff
@@ -0,0 +1,416 @@
+#!/usr/bin/env python3
+import argparse
+import dataclasses
+import os
+import re
+import shutil
+from os import listdir
+from os.path import isfile, join
+from subprocess import run as shrun, PIPE
+
+# Assumptions with the idea of creating conventions over configurations.
+
+# 1. Dockerfiles are self-contained, and they don't require any binary
+#    prebuilt on the host
+# 2. Dockerfiles are in the directory matching
+#    `openshift/ci-operator/knative-images/**/Dockerfile`
+#    or openshift/ci-operator/knative-test-images/**/Dockerfile
+# 3. The repository has a Makefile in the root and has some targets containing the keyword
+#    `test`
+
+# Usage:
+# ./generate_ci.py \
+#   --openshift-release-remove git@github.com:pierDipi/release.git # Remote to your fork
+
```

Comment:
@pierDipi I've read this but I still don't understand what this script does.

Can you write some explanation please? Then I can review

Reply:
it generates openshift/release configuration for configured repositories (repo variable) and supported branches.

Reply:
when we create a new branch in a midstream repository, the usual flow looks like this:

1. mirror upstream branch
2. apply openshift directory from main
3. go to openshift/release and add CI jobs and image mirroring for that branch

I'm automating 3 for _every_ midstream repository (except SO for reasons), and we will eventually try to automate 1 and 2 in this automation

Reply:
You could reuse a tool I wrote couple of months ago https://github.com/cardil/deviate

It's used for kn-plugin-event. Here's usage https://github.com/openshift-knative/kn-plugin-event/blob/main/openshift/release/update-to-head.sh#L9-L11

Reply:
@cardil for 1 and 2 right?

Reply:
we discussed your tool in our weekly call a while ago and it was one of the options we considered for 1 and 2.

Reply:
I like it, can you please showcase how we can use it in `openshift-knative/eventing-kafka-broker` or `openshift/knative-eventing`?

Reply:
because I see that the sync'ed branches are a bit different from what we have currently like:
upstream `release-1.5` -> midstream `release-v1.5`.

I'm sure we can add a feature to add the `v` if it's not there already.

Reply:
Those branch names and tags are configurable. It should be possible to use it after configuration.

I plan the Deviate tool could generate appropriate configs for openshift/release...